pkg/datapath/linux/arp/ping_linux.go

```diff
@@ -0,0 +1,187 @@
+// Copyright 2018-2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build linux
+
+package arp
+
+import (
+	"net"
+	"os"
+	"syscall"
+	"time"
+
+	"golang.org/x/sys/unix"
+)
+
+// protolARP is the uint16 EtherType representation of ARP (Address
+// Resolution Protocol, RFC 826). 0x0608 is htons(0x0806) as we need it
+// in network byte order
+const protoARP = 0x0608
+
+var _ net.PacketConn = &packetConn{}
+
+type packetConn struct {
+	iface *net.Interface
+	s     *arpSocket
+}
+
+func listen(iface *net.Interface) (*packetConn, error) {
+	us, err := unix.Socket(unix.AF_PACKET, unix.SOCK_RAW, protoARP)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := unix.SetNonblock(us, true); err != nil {
+		return nil, err
+	}
+
+	// if the socket is non-blocking, this returns a pollable resource
+	// which is needed for Set*Deadline() methods to work
+	f := os.NewFile(uintptr(us), "arp-socket")
+	rc, err := f.SyscallConn()
+	if err != nil {
+		return nil, err
+	}
+
+	as := &arpSocket{
+		f:  f,
+		rc: rc,
+	}
+
+	pc := &packetConn{
+		iface: iface,
+		s:     as,
+	}
+
+	return pc, nil
+}
+
+func (p *packetConn) ReadFrom(b []byte) (int, net.Addr, error) {
```

Comment:
Is this implementation bypasses the problem mentioned in https://github.com/cilium/arping/commit/5142c9527af5cbdcd939a9016c2ecf3bb431959b?

Reply:
I don't think that situation applies to the implementation in this PR as I use a non-blocking mode.